runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java

```diff
@@ -651,16 +652,12 @@ private void processElement(Map<W, W> windowToMergeResult, WindowedValue<InputT>
       // it but the local output watermark (also for this key) has not. After data is emitted and
       // the output watermark hold is released, the output watermark on this key will immediately
       // exceed the end of the window (otherwise we could see multiple ON_TIME outputs)
-      this.isEndOfWindow =
-          timerInternals.currentInputWatermarkTime().isAfter(window.maxTimestamp())
-              && outputWatermarkBeforeEOW;
+      this.isEndOfWindow = !timestamp.isBefore(window.maxTimestamp()) && outputWatermarkBeforeEOW;
 
       // The "GC time" is reached when the input watermark surpasses the end of the window
       // plus allowed lateness. After this, the window is expired and expunged.
       this.isGarbageCollection =
-          timerInternals
-              .currentInputWatermarkTime()
-              .isAfter(LateDataUtils.garbageCollectionTime(window, windowingStrategy));
+          !timestamp.isBefore(LateDataUtils.garbageCollectionTime(window, windowingStrategy));
```

Comment:
This one was tough. The problem here was, that previously input watermark moved at the end of each bundle. Each bundle also contained its own timers. But - because timers can generate new timers and these are processed in different bundle - the actual timestamp of timer must be taken into account. Otherwise a timer setup for some earlier time might trigger window garbage collection, because input watermark is already way ahead. I'm not quite sure of the other consequences though. Is it correct, that one bundle might generate multiple bundles, or is there a bundle atomicity requirement? On the other hand, runners that don't have concept of bundles (and have therefore by definition bundle of size 1) have to generate multiple bundles from single bundle, so that might be ok. Am I right?